Collect complexity statistics of solver queries

xbauch opened the pull request:
Introduces a functionality to track and collect the complexity of individual
solver queries. For SAT solvers we collect the number of clauses, literals, and
variables. The functionality matches solver queries to the SSA steps that
produced them. After running the solver, we produce a JSON report that matches
program line to GOTO instruction to SSA step to solver hardness.

At present the collection is command-line optional and only works with the
default SAT solver (minisat2). Extension to other SAT and SMT solvers should be
possible.

A simple regression test is included.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

xbauch commented on src/goto-checker/solver_factory.cpp:
```diff
@@ -11,6 +11,8 @@ Author: Daniel Kroening, Peter Schrammel
 
 #include "solver_factory.h"
 
+#include <sys/stat.h>
```
Will be removed.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+
+void solver_hardnesst::produce_report(const namespacet &ns)
+{
+  if(outfile.empty())
```
This behaviour is undocumented. It should be documented.

xbauch replied:
This should be part of the debate as to how to test this feature. Originally, I wanted only to print to a JSON file, but I did not know how to simply test the content of that file.

hannes-steffenhagen-diffblue commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -426,6 +426,11 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
 
+  if(cmdline.isset("solver-stats") && !cmdline.isset("paths"))
+  {
+    options.set_option("solver-stats", cmdline.get_value("solver-stats"));
```
The name of this option should reflect that it contains the name of a file to write output to. Maybe something as simple as `write-solver-stats-to`?

xbauch replied:
Renamed.

hannes-steffenhagen-diffblue commented on regression/cbmc/solver-hardness-simple/test.desc:
```diff
@@ -0,0 +1,12 @@
+^\s*\"GOTO_ID\": \d+,$
+^\s*\"Clauses\": \d+,$
+^\[main.assertion.\d+\] line \d+ assertion a \+ b \< 10: FAILURE$
+^VERIFICATION FAILED$
```
This is one of those cases where it might be worth having a separate test script that can actually parse the JSON instead; Not sure if it's worth doing, but we could use such a thing to make sure the output we get is actually valid JSON, and that it is of the structure we're expecting.

I'm also not quite convinced this covers all of the code of the stat collection

hannes-steffenhagen-diffblue commented on src/goto-checker/solver_factory.cpp:
```diff
@@ -11,6 +11,8 @@ Author: Daniel Kroening, Peter Schrammel
 
 #include "solver_factory.h"
 
+#include <sys/stat.h>
```
This header is for the posix stat syscalls. I don't think you need it here.

xbauch replied:
Removed.

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -424,31 +465,48 @@ void symex_target_equationt::convert_goto_instructions(
 
         step.cond_handle = decision_procedure.handle(step.cond_expr);
+
+        if(solver_hardness != nullptr)
```
Not the end of the world, but I'm usually not a huge fan of using `nullptr` to indicate optional things if we can avoid it

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -424,31 +465,48 @@ void symex_target_equationt::convert_goto_instructions(
-        step.output(mstream);
-        mstream << messaget::eom;
-      });
+    if(step.is_goto())
```
Please document why you need this additional check here

xbauch replied:
I don't, copy-pasted. Will be removed.

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -458,6 +516,7 @@ void symex_target_equationt::convert_assertions(
   if(number_of_assertions==0)
     return;
 
+  auto solver_hardness = decision_procedure.solver_hardness;
```
const

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -520,19 +601,34 @@ void symex_target_equationt::convert_assertions(
     }
   }
 
+  auto assertion_disjunction = disjunction(disjuncts);
```
const

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -520,19 +601,34 @@ void symex_target_equationt::convert_assertions(
+  decision_proceduret &dec_proc,
+  const namespacet &ns)
 {
+  auto solver_hardness = dec_proc.solver_hardness;
```
const

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -520,19 +601,34 @@ void symex_target_equationt::convert_assertions(
   for(auto &step : SSA_steps)
     if(!step.ignore)
     {
+      and_exprt::operandst conjuncts;
```
what's happening here?

xbauch replied:
There is a number of solver calls for a single SSA step, each representing an equality between function argument and parameter. I wanted a single expression for all there equalities.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+  current_stats.ssa_expression = ssa_expression;
+  current_stats.pc = pc;
+  hardness_stats.push_back(current_stats);
+  current_stats.clear();
```
What's with the clear?

xbauch replied:
Well the solver always updates the statistics into `current_stats` and it's up to the symex to match them to an SSA step. After the match, we clean the `current_stats` before the solver starts updating again.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+  std::sort(
+    hardness_stats.begin(),
+    hardness_stats.end(),
+    [](const hardness_statst &l_stats, const hardness_statst &r_stats) {
```
I think this should probably be moved out into its own function for clarity, something like 

```
static bool order_by_instruction_source_location(const hardness_statst& l_stats, const hardness_statst &r_stats);
``` 

```
std::sort(hardness_stats.begin(), hardness_stats.end(), order_by_instruction_source_location);
```

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+    // steps per GOTO instruction. The following JSON contains a unique
+    // identifier for each one.
+    hardness_stats_json["GOTO_ID"] = json_numbert{
+      std::to_string(reinterpret_cast<std::uintptr_t>(&(*stat.pc)))};
```
This should be fine for machine processing, but we should still consider at some point if we shouldn't have more predictable "identifiers".

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+    // to a particular GOTO instruction, since there may be a number of SSA
+    // steps per GOTO instruction. The following JSON contains a unique
+    // identifier for each one.
+    hardness_stats_json["GOTO_ID"] = json_numbert{
```
I'm still baffled that a json number is a string

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+      json_numbert{std::to_string(stat.sat_hardness.clauses)};
+    sat_hardness_json["Literals"] =
+      json_numbert{std::to_string(stat.sat_hardness.literals)};
+    sat_hardness_json["Variables"] =
```
super minor point, but in my experience it's more common to use `camelCase` or `snake_case` for identifiers in JSON; This isn't at all required, but it might trip some people up a bit to see something else.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.h:
```diff
@@ -0,0 +1,166 @@
+  assertion_statst assertion_stats;
+
+  // A minor modification of \ref goto_programt::output_instruction
+  std::string goto_instruction2string(
```
This should be static

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,324 @@
+std::string solver_hardnesst::goto_instruction2string(
+  goto_programt::const_targett pc,
+  const namespacet &ns) const
+{
```
I'm not completely sold on the idea of repeating so much of the printing logic here. I get the existing printing functionality didn't quite meet our purposes here, but it still feels a bit "off" to repeat this ad-hoc over here hidden in a private function somewhere.

hannes-steffenhagen-diffblue reviewed (commented):
I think the overall direction here is OK, but there's some bits that might need cleanup or a rethink

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -379,13 +404,21 @@ void symex_target_equationt::convert_guards(
+      if(solver_hardness != nullptr)
+      {
+        solver_hardness->register_ssa(
+          from_expr(ns, ID_C, step.guard), step.source.pc);
```
🚫 If you need a namespace then pass it through solver_hardness and don't thread it through symex_target_equation. Besides, from_expr is pretty much deprecated.

xbauch replied:
Done. Went for `format` instead. Prints `main::1::a` instead of `a`, but ok.

peterschrammel commented on src/solvers/solver_hardness.cpp:
```diff
@@ -0,0 +1,329 @@
+#include <util/string2int.h>
+
+void solver_hardnesst::register_ssa(
+  const std::string &ssa_expression,
```
Take an expr here

xbauch replied:
Done.

peterschrammel commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -45,7 +45,7 @@ class optionst;
   OPT_REACHABILITY_SLICER \
   "(debug-level):(no-propagation)(no-simplify-if)" \
-  "(document-subgoals)(outfile):(test-preprocessor)" \
+  "(document-subgoals)(outfile):(write-solver-stats-to):(test-preprocessor)"  \
```
put on separate lines

xbauch replied:
Done.

xbauch commented:
@hannes-steffenhagen-diffblue small rework in the second commit.

martin-cs commented:
Forgive the likely daft question but ... why?  The cost per SSA instruction is generally pretty straight-forward and there are only really a few costly operations -- you could just look for those.  Maybe I'm just not getting this...

xbauch commented:
@martin-cs I guess you could call it a feature request. At the beginning I had some filtering for non-trivial steps, but it was suggested that people could do it on their own, once they have the JSON. I think the big-picture motivation is to overlay the statistics in some IDE so that you can track how hard each line of code is. Code -> GOTO -> SSA -> clauses. That way you can optimise your code (for the solver) on whichever level you choose.

martin-cs commented:
OK... but... 1. clauses / variables are a measure of size, not difficulty, 2. the amount used depends on the context not just the assignment, 3. you can approximate expensive instructions much more simply, 4. the really big wins are from reorganisations of the code (see the N different ways of doing recursion) not individual instructions, 5. the inbuilt solver is *very* naive and you won't really be able to track similar stats for other solvers.

I'm not going to oppose this, esp. if this is something you are doing for a specific customer request but ... I think there may be a better way of achieving your high-level goal and I worry that this may not actually solve your problem.

hannes-steffenhagen-diffblue commented on src/goto-checker/incremental_goto_checker.h:
```diff
@@ -83,6 +87,8 @@ class incremental_goto_checkert
   ui_message_handlert &ui_message_handler;
   messaget log;
+
+  std::shared_ptr<solver_hardnesst> solver_hardness;
```
I think you explained this at some point but I still don't quite get why this is shared rather than unique. Do all the things that want a handle on this thing have different/non-hierarchical lifetimes?

hannes-steffenhagen-diffblue reviewed (approved):
Codewise, tis is probably fine modulo some smaller concerns, but this should probably not be merged until we're sure we're actually solving the right problem here.

martin-cs reviewed (commented):
Github seems to want a review from me for this so...

I think there are likely easier ways of achieving most of the information that you can get like this.  I don't think that information, regardless how it is computed, is really what you need to solve actual end-user problems.  I am happy to help with what I know about solver hardness / to discuss this if it would help.

But, I'm not going to block this, so the review is "comment" not "request changes".

karkhaz reviewed (approved):
Thanks, I believe this will be useful. Happy for this to go in!

@martin-cs thanks for your comments–they're very valid, and I do hope that people don't read more into these numbers than what they really mean. We have a very specific use-case for them, which I'm unable to share unfortunately...

peterschrammel commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -46,6 +46,7 @@ class optionst;
   OPT_REACHABILITY_SLICER \
   "(debug-level):(no-propagation)(no-simplify-if)" \
   "(document-subgoals)(outfile):(test-preprocessor)" \
+  "(write-solver-stats-to):"  \
```
Define an `OPT_...`

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1106,6 +1112,8 @@ void cbmc_parse_optionst::help()
     HELP_FLUSH
     " --verbosity #                verbosity level\n"
     HELP_TIMESTAMP
+    " --write-solver-stats-to json-file\n"
```
Define a `HELP_...`

petr-bauch committed:
Track hardness of solver queries

Introduces a functionality to track and collect the complexity of individual
solver queries. For SAT solvers we collect the number of clauses, literals, and
variables. The functionality matches solver queries to the SSA steps that
produced them. After running the solver, we produce a JSON report that matches
program line to GOTO instruction to SSA step to solver hardness.

At present the collection is command-line optional and only works with the
default SAT solver (minisat2). Extension to other SAT and SMT solvers should be
possible.

A simple regression test is included.

Note: the solver-hardness structure lives in the minisat2 and is accessible via
the interface of hardness-collector (which minisat2 inherits from). Hence to
find out if symex should collect the statistics one must try-cast `prop` to
`hardness-collector`. `prop` is part of the `solvert` but not directly
accessible from `decision-procedure`. So we also had to extend the interface
that initialises the SSA conversion (to include `prop` as well).

xbauch commented:
@peterschrammel I'm not sure this is exactly what you meant; I didn't know how to get the `prop` (since minisat is a `prop`) out of `decision-procedure` so I extended the conversion interface for symex-target-equation.

peterschrammel commented on src/goto-symex/symex_target_equation.h:
```diff
@@ -178,60 +181,93 @@ class symex_target_equationt:public symex_targett
 
+  /// \copydoc convert(decision_proceduret &)
+  /// \param prop: A handle to a prop interface to query the solver hardness
+  void convert(decision_proceduret &decision_procedure, propt &prop);
```
🚫 None of these signature changes are necessary.

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -583,10 +714,18 @@ void symex_target_equationt::convert_io(
           step.converted_io_args.push_back(symbol);
         }
       }
+      if(solver_hardness != nullptr)
```
Check whether decision_procedure implements hardness_collectort. If yes, then do the necessary.
You may want to write a little helper method to wrap that logic to avoid repetition throughout the conversion functions.

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -319,27 +321,52 @@ void symex_target_equationt::constraint(
+  }();
+
+  convert_without_assertions(decision_procedure, solver_hardness);
+  convert_assertions(decision_procedure, solver_hardness); //, solver_hardness);
```
remove commented code

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -319,27 +321,52 @@ void symex_target_equationt::constraint(
 {
-  convert_without_assertions(decision_procedure);
-  convert_assertions(decision_procedure);
+  convert_without_assertions(decision_procedure, nullptr);
```
I don't want to pass in a `nullptr` when I don't need the feature.

peterschrammel commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -319,27 +321,52 @@ void symex_target_equationt::constraint(
-  convert_io(decision_procedure);
-  convert_constraints(decision_procedure);
+  auto solver_hardness = [&prop]() -> solver_hardnesst * {
+    if(auto hardness_collector = dynamic_cast<hardness_collectort *>(&prop))
```
🚫 Make `prop_conv_solvert` implement `hardness_collectort` and forward the calls. `propt` MUST not be exposed here.

peterschrammel commented on src/goto-checker/single_loop_incremental_symex_checker.cpp:
```diff
@@ -89,15 +89,15 @@ operator()(propertiest &properties)
         log.status() << "converting SSA" << messaget::eom;
         equation.convert_without_assertions(
-          property_decider.get_decision_procedure());
+          property_decider.get_decision_procedure(), nullptr);
```
🚫

peterschrammel commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -155,3 +166,19 @@ multi_path_symex_checkert::localize_fault(const irep_idt &property_id) const
+  {
+    if(
+      auto hardness_collector =
+        dynamic_cast<hardness_collectort *>(&property_decider.get_prop()))
```
See below. `propt` must not be exposed. Forward to `prop_conv_solvert` which is a `decision_proceduret`.

peterschrammel commented on src/goto-checker/goto_symex_property_decider.cpp:
```diff
@@ -110,6 +110,11 @@ goto_symex_property_decidert::get_decision_procedure() const
   return solver->decision_procedure();
 }
 
+propt &goto_symex_property_decidert::get_prop() const
```
Unnecessary.

peterschrammel commented on src/goto-checker/bmc_util.h:
```diff
@@ -36,7 +36,7 @@ class ui_message_handlert;
 void convert_symex_target_equation(
   symex_target_equationt &,
-  decision_proceduret &,
+  goto_symex_property_decidert &,
```
🚫

peterschrammel commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -149,14 +150,22 @@ void output_graphml(
 
   // convert SSA
-  equation.convert(decision_procedure);
+  if(dynamic_cast<boolbvt *>(&property_decider.get_decision_procedure()))
```
🚫  unnecessary

peterschrammel reviewed (changes requested):
From my conversion with @xbauch, the forwarding from `propt` to `prop_conv_solvert` was missed. Doing that removes most of the ugly plumbing from this PR.

hannes-steffenhagen-diffblue commented:
Closing in favour of https://github.com/diffblue/cbmc/pull/5266
